### PR TITLE
logical_range_filter: add more query-log

### PR DESCRIPTION
**Pull request by kenhys**

### CODE REVIEW

File: test/command/suite/sharding/logical_range_filter/columns/stage/filtered/window_function/day/max.expected [hunk 1]
```diff
@@ -86,5 +86,754 @@ logical_range_filter Logs   --max '2017/03/16 08:22:00'   --max_border include
...
+#:000000000000000 filter(0): between(#<accessor _key(anonymous).timestamp(Logs_20170316)>, "2017/03/02 00:00:00", "include", "2017/03/02 01:00:00", "exclude")
+#:000000000000000 filter(0): between(#<accessor _key(anonymous).timestamp(Logs_20170316)>, "2017/03/02 01:00:00", "include", "2017/03/02 02:00:00", "exclude")
+#:000000000000000 filter(0): between(#<accessor _key(anonymous).timestamp(Logs_20170316)>, "2017/03/02 02:00:00", "include", "2017/03/02 03:00:00", "exclude")
+#:000000000000000 filter(0): between(#<accessor _key(anonymous).timestamp(Logs_20170316)>, "2017/03/02 03:00:00", "include", "2017/03/02 04:00:00", "exclude")
+#:000000000000000 filter(0): between(#<accessor _key(anonymous).timestamp(Logs_20170316)>, "2017/03/02 04:00:00", "include", "2017/03/02 05:00:00", "exclude")
+#:000000000000000 filter(0): between(#<accessor _key(anonymous).timestamp(Logs_20170316)>, "2017/03/02 05:00:00", "include", "2017/03/02 06:00:00", "exclude")
+#:000000000000000 filter(0): between(#<accessor _key(anonymous).timestamp(Logs_20170316)>, "2017/03/02 06:00:00", "include", "2017/03/02 07:00:00", "exclude")
+#:0000000000000
```
Comment (komainu8) [inline test/command/suite/sharding/logical_range_filter/columns/stage/filtered/window_function/day/max.expected]: In my opinion, too many output logs.
Therefore, we are afraid that this output influence performance.

Is not there an influence in performance?

Comment (kenhys) [reply test/command/suite/sharding/logical_range_filter/columns/stage/filtered/window_function/day/max.expected]: Oh, It may affects.

Comment (kenhys) [reply test/command/suite/sharding/logical_range_filter/columns/stage/filtered/window_function/day/max.expected]: It is generated by table.select(expression) in filter_by_range.

**Comment by kenhys** [comment]
@komainu8 could you check again?

File: lib/expr.c [hunk 2]
```diff
@@ -5029,14 +5029,17 @@ grn_table_select(grn_ctx *ctx, grn_obj *table, grn_obj *expr,
...
-                        ctx,
-                        &condition_inspect_buffer,
-                        si,
-                        e));
+        const char *condition = grn_table_select_inspect_condition(ctx,
+                                                                   &condition_inspect_buffer,
+                                                                   si,
+                                                                   e);
+        if (grn_strncasecmp(condition, ": between", 9) != 0 ||
+            (grn_strncasecmp(condition, ": between", 9) == 0 && grn_table_size(ctx, data.res) > 0)) {
```
Comment (komainu8) [inline lib/expr.c]: Umm...
I think `filter(0)` is also important information.
However, I have no good idea against this modify.

Could you directly ask kou to review, please?

Comment (kenhys) [reply lib/expr.c]: @komainu8 
Ok, I'll do so.  Thank you for review!

**Comment by kenhys** [comment]
@kou 
Could you review this PR please?
One concern is how to reduce a bit redundant query logs such as `filter(0): between`.

File: test/command/suite/sharding/logical_range_filter/use_range_index/yes/filter/sub_filter.expected [hunk 3]
```diff
@@ -61,3 +61,5 @@ logical_range_filter Logs timestamp   --use_range_index yes   --filter 'sub_filt
   ]
 ]
 #|e| selector only proc can't be called: <sub_filter>
+#>logical_range_filter --filter "sub_filter(action.errors_action, \"level == \\\"critical\\\"\")" --logical_table "Logs" --shard_key "timestamp" --use_range_index "yes"
+#<000000000000000 rc=-38
```
Comment (kou) [inline test/command/suite/sharding/logical_range_filter/use_range_index/yes/filter/sub_filter.expected]: Do we need query log for this test?

File: test/command/suite/sharding/logical_range_filter/columns/stage/filtered/window_function/month/time_classify_mix.expected [hunk 4]
```diff
@@ -80,5 +80,27 @@ logical_range_filter Logs   --shard_key timestamp   --columns[day].stage filtere
...
+#:000000000000000 filter(11): Logs_201703.price greater_equal 300
+#:000000000000000 filter(3): between(#<accessor _key(anonymous).timestamp(Logs_201703)>, "2017/03/15 00:00:00", "include", "2017/03/16 00:00:00", "exclude")
+#:000000000000000 columns[day](3)
+#:000000000000000 columns[five_minute](3)
+#:000000000000000 columns[n_records_per_day](3)
+#:000000000000000 columns[n_records_per_five_minute](3)
+#:000000000000000 filter(0): 2147483651.n_records_per_day greater_equal 4
+#:000000000000000 filter(3): 2147483651.n_records_per_five_minute greater_equal 3
+#:000000000000000 post_filter(3): Logs_201703.n_records_per_day >= 4 || n_records_per_five_minute >= 3
```
Comment (kou) [inline test/command/suite/sharding/logical_range_filter/columns/stage/filtered/window_function/month/time_classify_mix.expected]: It this expected?

File: plugins/sharding/logical_range_filter.rb [hunk 5]
```diff
@@ -970,10 +972,14 @@ def filter_by_range(range_index, min, min_border, max, max_border)
                     n_matched_records = index_cursor.select(result_set, options)
                   end
                 end
+                query_logger.log(:size, ":",
+                                 "filter(#{n_matched_records}): #{@shard.table_name}.#{@filter}")
```
Comment (kou) [inline plugins/sharding/logical_range_filter.rb]: `filter(N)[TABLE_NAME]: FILTER` is better.

File: test/command/suite/sharding/logical_range_filter/invalid/no_shard.expected [hunk 6]
```diff
@@ -9,3 +9,5 @@ logical_range_filter Logs timestamp
     "[logical_range_filter] no shard exists: logical_table: <Logs>: shard_key: <timestamp>"
   ]
 ]
+#>logical_range_filter --logical_table "Logs" --shard_key "timestamp"
+#<000000000000000 rc=-22
```
Comment (kou) [inline test/command/suite/sharding/logical_range_filter/invalid/no_shard.expected]: Do we need this?

File: plugins/sharding/logical_range_filter.rb [hunk 7]
```diff
@@ -1057,7 +1063,11 @@ def filter_table
         def apply_post_filter(table)
           create_expression(table) do |expression|
             expression.parse(@post_filter)
-            table.select(expression)
+            filtered_result = table.select(expression)
+            query_logger.log(:size,
+                           ":",
+                           "post_filter(#{filtered_result.size}): #{@shard.table_name}.#{@post_filter}")
```
[the remark this answers is not in the document]
Comment (kou) [reply plugins/sharding/logical_range_filter.rb]: Is `.#{@post_filter}` needed? Is it logged by `filter()`?

[the remark this answers is not in the document]
Comment (kenhys) [reply plugins/sharding/logical_range_filter.rb]: Ah, `filter()` covers it.

File: plugins/sharding/logical_range_filter.rb [hunk 8]
```diff
@@ -1112,6 +1122,12 @@ def sort_result_set(result_set)
           if @context.current_limit > 0
             @context.current_limit -= sorted_result_set.size
           end
+          keys =  sort_keys.collect do |key_order_pair|
```
[the remark this answers is not in the document]
Comment (kenhys) [reply plugins/sharding/logical_range_filter.rb]: I'll fix.

File: plugins/sharding/logical_range_filter.rb [hunk 9]
```diff
@@ -1112,6 +1122,12 @@ def sort_result_set(result_set)
...
+            key_order_pair[:key]
```
[the remark this answers is not in the document]
Comment (kenhys) [reply plugins/sharding/logical_range_filter.rb]: There is no strong reason. I'll fix it.

File: test/command/suite/sharding/logical_range_filter/sort_keys/multiple.expected [hunk 10]
```diff
@@ -123,5 +123,17 @@ logical_range_filter Logs   --shard_key timestamp   --columns[hour2].stage initi
 #|d| [logical_range_filter][select] <Logs_20170315>: dynamic columns are used
 #|d| [logical_range_filter][select] <Logs_20170316>: dynamic columns are used
 #|d| [logical_range_filter][select] <Logs_20170317>: dynamic columns are used
+#>logical_range_filter --columns[hour2].flags "COLUMN_SCALAR" --columns[hour2].stage "initial" --columns[hour2].type "Time" --columns[hour2].value "time_classify_hour(timestamp, 2)" --logical_table "Logs" --output_columns "_id,hour2,price" --shard_key "timestamp" --sort_keys "hour2,-price"
+#:000000000000000 filter(3): true
+#:000000000000000 columns[hour2](3)
+#:000000000000000 sort(3): Logs_20170315.hour2,price
```
[the remark this answers is not in the document]
Comment (kenhys) [reply test/command/suite/sharding/logical_range_filter/sort_keys/multiple.expected]: I'll fix it.

File: test/command/suite/sharding/logical_range_filter/use_range_index/yes/filter/sub_filter.expected (hunk 3 above)
Comment (kenhys) [reply test/command/suite/sharding/logical_range_filter/use_range_index/yes/filter/sub_filter.expected]: Ah, it is needless.

File: test/command/suite/sharding/logical_range_filter/columns/stage/filtered/window_function/month/time_classify_mix.expected (hunk 4 above)
Comment (kenhys) [reply test/command/suite/sharding/logical_range_filter/columns/stage/filtered/window_function/month/time_classify_mix.expected]: Already coverted by `filter()`. I'll fix.

File: plugins/sharding/logical_range_filter.rb (hunk 5 above)
Comment (kenhys) [reply plugins/sharding/logical_range_filter.rb]: I'll fix it.

File: test/command/suite/sharding/logical_range_filter/invalid/no_shard.expected (hunk 6 above)
Comment (kenhys) [reply test/command/suite/sharding/logical_range_filter/invalid/no_shard.expected]: It is needless. I'll fix it.

**Comment by kenhys** [comment]
@kou I've removed `filter():` omssion, and pushed some commit which is based on feedback!

File: plugins/sharding/logical_range_filter.rb (hunk 5 above)
Comment (kou) [reply plugins/sharding/logical_range_filter.rb]: Really? Is `"filter(#{n_matched_records}): #{@shard.table_name}.#{@filter}"` expected?

File: plugins/sharding/logical_range_filter.rb [hunk 11]
```diff
@@ -1112,6 +1118,13 @@ def sort_result_set(result_set)
           if @context.current_limit > 0
             @context.current_limit -= sorted_result_set.size
           end
+          key_order_pairs = ""
+          sort_keys.each do |pair|
+            key_order_pairs << "key: #{pair[:key]} order: #{pair[:order]}"
+          end
```
Comment (kou) [inline plugins/sharding/logical_range_filter.rb]: `KEY` for asc and `-KEY` for desc is better. We use it in other places.

File: plugins/sharding/logical_range_filter.rb (hunk 5 above)
Comment (kenhys) [reply plugins/sharding/logical_range_filter.rb]: Oops!

File: plugins/sharding/logical_range_filter.rb (hunk 11 above)
Comment (kenhys) [reply plugins/sharding/logical_range_filter.rb]: Thanks for review.

**Comment by kenhys** [comment]
@kou I've fixed missing logging style and unify to KEY/-KEY style.

**Comment by kou** [comment]
Merged with some modifications.

**Pull request merged**